### ISSUE TITLE
2 fold adjudicated as a draw

### ISSUE THREAD

**Issue by Sopel97** [nodchip/Stockfish#298]
https://github.com/nodchip/Stockfish/blob/74774c36e19d43f56d533e16d778582239de8934/src/learn/gensfen.cpp#L493

is_draw returns true on 2 fold, but this can happen in won games. It makes sense to use in a search, but doesn't make sense to adjudicate games.

**Comment by fsmosca** [comment]
Does gensfen always generate random moves? If so then `pos.is_draw()` is just random and it is better not to use it.

I thought about an option where at certain ply limit, moves will be based on search - no more random. In this case, `pos.is_draw()` can be safely applied.

**Comment by Sopel97** [comment]
gensfen does random moves but only up to some ply (see the docs, I don't remember exactly what governs it, but random moves stop pretty quick)

**Comment by fsmosca** [comment]
According to the doc if `random_multi_pv` is not specified then all moves will be random except when there is only one legal move. If specified, then a random move from multi_pv value will be selected. If user sets random_multi_pv to 1 then this is not random as multipv 1 is a normal search. Above 1 it is random except when there is only one legal move.

Generally there is no issue with 2-fold repetition, assuming that moves are coming from normal search - not from some random moves. In normal search if the engine cannot find a better move it will go for repetition if there is one. The use of pos.is_draw(ply) is good because we terminate the game early as we are not waiting for the engine to execute the 3-fold repetition.

**Comment by Sopel97** [comment]
> `random_move_minply` - the minimal ply at which a random move may be executed instead of a move chosen by search. Default: 1.
> 
> `random_move_maxply` - the maximal ply at which a random move may be executed instead of a move chosen by search. Default: 24.
> 
> `random_move_count` - maximum number of random moves in a single self-play game. Default: 5.

**Comment by fsmosca** [comment]
This `random_move_count` looks interesting. Does it mean that after 5 random moves are already in the game, no more random moves will be executed. If that is so then pos.is_draw(ply) can be used then.

**Comment by Sopel97** [comment]
yes that is true

**Comment by fsmosca** [comment]
To erase any doubt I have a [revision to adjudicate the game by 3-fold repetition](https://github.com/fsmosca/Stockfish/commits/3fold_rep_termination).

Tested to work with the following command.
```
gensfen random_multi_pv 3 random_multi_pv_diff 200 set_recommended_uci_options book noob_3moves.epd depth 5 loop 10000 ..."
```

Sample games:

[Event "?"]
[Site "?"]
[Date "?"]
[Round "?"]
[White "?"]
[Black "?"]
[Result "1/2-1/2"]
[FEN "rnbqkbnr/p1p2ppp/1p2p3/3p4/P3P3/N7/1PPP1PPP/R1BQKBNR w KQkq - 0 1"]
[SetUp "1"]

1.Bd3 Nf6 2.exd5 Nxd5 3.Nf3 Nf4 4.O-O Nxd3 5.cxd3 Nc6 6.d4 Be7 7.d3 O-O 8.Nc2 Bb7 9.Bf4 a5 10.Re1 Nb4 11.Nxb4 Bxf3 12.gxf3 Bxb4 13.Re4 Qd7 14.Rc1 c6 15.Bd2 Rfd8 16.Bxb4 axb4 17.f4 Ra5 18.Rc4 Qd6 19.b3 Rd5 20.Qd2 f5 21.Re1 Rxd4 22.Rxd4 Qxd4 23.Rxe6 Qc3 24.Qxc3 bxc3 25.Rxc6 Rxd3 26.Rc7 h6 27.Kf1 g5 28.fxg5 hxg5 29.Ke2 Rd2+ 30.Ke1 Rc2 31.b4 Rb2 32.b5 Ra2 33.Rxc3 Rxa4 34.Rc6 Kf7 35.Rxb6 Rb4 36.Kd2 Rb2+ 37.Kc3 Rxf2 38.Rh6 Re2 39.b6 Re8 40.Rc6 f4 41.Rc5 Kf6 42.Rc6+ Kf7 43.Rc5 Kf6 44.Rc6+ Kf7 1/2-1/2

[Event "?"]
[Site "?"]
[Date "?"]
[Round "?"]
[White "?"]
[Black "?"]
[Result "1/2-1/2"]
[FEN "rnbqkbnr/pp2ppp1/2p4p/3p4/8/2P3P1/PPQPPP1P/RNB1KBNR w KQkq - 0 1"]
[SetUp "1"]

1.d4 Nf6 2.Nf3 e6 3.Bg2 Bd6 4.Bd2 Nbd7 5.O-O b6 6.Be3 Ng4 7.Bf4 Bxf4 8.gxf4 Bb7 9.Nbd2 O-O 10.h3 Ngf6 11.Ne5 c5 12.Qa4 Qc7 13.e3 Nxe5 14.fxe5 Nd7 15.Qd1 a5 16.a4 f6 17.f4 Ba6 18.Rf2 fxe5 19.Qg4 exd4 20.cxd4 Qc6 21.Bf3 Nf6 22.Qg3 cxd4 23.Rg2 Ra7 24.exd4 Rc7 25.Kh2 Bd3 26.Ra3 Bf5 27.Rc3 Qd7 28.Qxg7+ Qxg7 29.Rxc7 Rf7 30.Rxg7+ Kxg7 31.Rc6 Rd7 32.Rxb6 Ne4 33.Nf1 Kf6 34.Ne3 Kg6 35.Bg4 Bxg4 36.hxg4 Rd6 37.Rb5 Ra6 38.f5+ exf5 39.gxf5+ Kg5 40.Kh3 h5 41.Rxd5 Rd6 42.Rxa5 Rxd4 43.f6+ Kxf6 44.Kh4 Kg6 45.Ra6+ Nf6+ 46.Kg3 Kg5 47.Ra5+ Kg6 48.Ra7 Rb4 49.a5 Rxb2 50.a6 Ra2 51.Ra8 Kg5 52.Kf3 Ra3 53.a7 h4 54.Kg2 Rxe3 55.Rg8+ Nxg8 56.a8=Q Rg3+ 57.Kh1 Rh3+ 58.Kg2 Rg3+ 59.Kh1 Rh3+ 60.Kg2 Rg3+ 1/2-1/2

[Event "?"]
[Site "?"]
[Date "?"]
[Round "?"]
[White "?"]
[Black "?"]
[Result "1/2-1/2"]
[FEN "rnbqkbnr/1pp1ppp1/7p/p2p4/P7/3PP3/1PP2PPP/RNBQKBNR w KQkq - 0 1"]
[SetUp "1"]

1.Nf3 e6 2.d4 Nd7 3.Bd3 Bd6 4.O-O e5 5.dxe5 Bxe5 6.Nc3 Ngf6 7.Bf5 c6 8.Nxe5 Nxe5 9.Bxc8 Qxc8 10.b3 Qe6 11.Ba3 Ned7 12.Qd4 b6 13.Rad1 O-O-O 14.Bb2 Kb7 15.Ne2 Rhe8 16.Nf4 Qe7 17.Qc3 Qb4 18.Qxb4 axb4 19.Rfe1 Ra8 20.Nd3 c5 21.Nf4 Kc6 22.Ne2 g5 23.Ng3 h5 24.h3 h4 25.Nf5 Ne4 26.Ng7 Rh8 27.Nf5 Rhe8 28.Ng7 Red8 29.Nf5 Re8 1/2-1/2

[Event "?"]
[Site "?"]
[Date "?"]
[Round "?"]
[White "?"]
[Black "?"]
[Result "1/2-1/2"]
[FEN "rnbqkb1r/p1ppppp1/1p3n1p/8/7P/4PN2/PPPP1PP1/RNBQKB1R w KQkq - 0 1"]
[SetUp "1"]

1.Nc3 e6 2.d4 Bb7 3.Bd3 c5 4.O-O Nc6 5.a3 Be7 6.dxc5 bxc5 7.e4 d5 8.exd5 exd5 9.Bf4 Nh5 10.Bd2 O-O 11.b4 cxb4 12.axb4 Nxb4 13.Nd4 Nf6 14.Ncb5 a6 15.Bxb4 Bxb4 16.c3 Be7 17.Na3 Ne4 18.Bxe4 dxe4 19.Nc4 Rc8 20.Rb1 Qc7 21.g3 Rb8 22.Ne3 Rfd8 23.Qb3 a5 24.Nb5 Qc8 25.Nd4 Ba6 26.Qxb8 Qxb8 27.Rxb8 Rxb8 28.Ra1 Rc8 29.Rxa5 Bd3 30.Ra7 Kf8 31.Kg2 Bc5 32.Rb7 h5 33.Kg1 Ra8 34.Kg2 g6 35.Rc7 Bd6 36.Rd7 Be5 37.Nd5 Kg8 38.Re7 Bg7 39.Rd7 Bf8 40.Nb4 Bg7 41.Nxd3 exd3 42.Nb5 Bf8 43.Rxd3 Kg7 44.Rd7 Ra2 45.Kf3 Bc5 46.Nd4 Kf6 47.Rb7 Ra1 48.Rc7 Bd6 49.Rc8 Ke7 50.Rc6 Kd7 51.Ke4 Rc1 52.Ra6 Bc5 53.Kd3 Rf1 54.Ra2 Bb6 55.Rb2 Bxd4 56.Kxd4 Kd6 57.Ra2 Rd1+ 58.Kc4 Ke5 59.Kc5 Rd7 60.c4 Rc7+ 61.Kb4 Rb7+ 62.Kc3 Kd6 63.Kd4 Rd7 64.Ra6+ Ke7+ 65.Kc5 Rd8 66.Kb5 Rd2 67.f4 Rb2+ 68.Kc6 Rb3 69.Kd5 Rd3+ 70.Ke5 Re3+ 71.Kd4 Rxg3 72.c5 Rh3 73.Kd5 Rd3+ 74.Ke5 Re3+ 75.Kd4 Rh3 76.Ra7+ Ke6 77.c6 Rxh4 78.c7 Kd7 79.c8=N+ Kxc8 80.Rxf7 Rh1 81.Ke5 Kd8 82.Rh7 h4 83.Kf6 h3 84.Kxg6 h2 85.Kf6 Kc8 86.Kf5 Kd8 87.Kf6 Kc8 88.Kf5 Kd8 89.Kf6 1/2-1/2

[Event "?"]
[Site "?"]
[Date "?"]
[Round "?"]
[White "?"]
[Black "?"]
[Result "1/2-1/2"]
[FEN "rnbqkb1r/ppp2ppp/3ppn2/8/1P6/P4N2/2PPPPPP/RNBQKB1R w KQkq - 0 1"]
[SetUp "1"]

1.e4 Nxe4 2.Qe2 d5 3.Nc3 Nxc3 4.dxc3 Qf6 5.Bd2 Nd7 6.O-O-O c5 7.Kb1 c4 8.Re1 Qg6 9.Nh4 Qf6 10.Nf3 h5 11.Kb2 Bd6 12.Kb1 O-O 13.Ng5 Qg6 14.g4 hxg4 15.h3 Be7 16.Qxg4 a5 17.Nxe6 Qxg4 18.hxg4 fxe6 19.Rxe6 Bd8 20.Rd6 axb4 21.cxb4 Rxa3 22.Rxd5 c3 23.Be3 Kf7 24.Rb5 Ke8 25.Bc4 Be7 26.Rh7 g6 27.g5 Ra4 28.Rg7 Kd8 29.Rxg6 Rxb4+ 30.Rxb4 Bxb4 31.Re6 Bc5 32.g6 Rg8 33.Bxc5 Nxc5 34.Re5 Nd7 35.Bxg8 Nxe5 36.g7 Ng4 37.Ba2 Nf6 38.g8=R+ Nxg8 39.Bxg8 Ke7 40.Bh7 Kd6 41.Bd3 Be6 42.f4 Bd5 43.Kc1 b6 44.Kd1 Kc5 45.Ke2 Kd4 46.f5 Ke5 47.f6 Kxf6 48.Ke3 Ke5 49.Bb5 Be4 50.Bd3 Bc6 51.Bg6 b5 52.Kd3 Kf6 53.Bh7 Kg7 54.Bf5 Kf6 55.Bh7 Kg7 56.Bf5 Kf6 57.Bh7 1/2-1/2

**Comment by Sopel97** [comment]
make a pr :)

**Comment by Sopel97** [comment]
Fixed by #317
